File: 2020/python/day5.py
from utils import file_read_lines
from functools import reduce

lines = file_read_lines('../_inputs/day5.txt')


def run_part1():
  print(max(calculate_seats_id(boarding_pass) for boarding_pass in lines))


def run_part2():
  seats_ids = sorted((calculate_seats_id(boarding_pass) for boarding_pass in lines))

  last_id = seats_ids[0]

  for seat_id in seats_ids:
    if seat_id - last_id > 1:
      print(last_id + 1)
      break

    last_id = seat_id


def calculate_seats_id(boarding_pass):
  row_config = boarding_pass[:7]
  column_config = boarding_pass[7:]

  row = calculate(row_config, "B")
  column = calculate(column_config, "R")

  return row * 8 + column


def calculate(chars, truthy_char):
  return sum(2 ** index for index, char in enumerate(reversed(chars)) if char == truthy_char)

File: 2020/rust/src/day5.rs
use std::ops::RangeInclusive;
use itertools::sorted;

use super::utils;

pub fn run_part1() -> i32 {
  let result = calculate_seats_ids().max().unwrap();

  println!("{}", result);

  result
}

pub fn run_part2() -> i32 {
  let ids: Vec<i32> = sorted(calculate_seats_ids()).collect();

  let mut result = ids[0];

  for id in ids {
    if id - result > 1 {
      result += 1;
      break;
    }

    result = id;
  }

  println!("{}", result);

  result
}

fn calculate_seats_ids() -> impl Iterator<Item = i32> {
  utils::file::read_lines("../_inputs/day5.txt")
    .map(|line| {
      let (row_config, column_config) = line.split_at(7);

      let row = row_config
        .chars()
        .fold(RangeInclusive::new(0, 127), |range, letter| {
          let half = (range.end() + 1 - range.start()) / 2;

          match letter {
            'F' => RangeInclusive::new(*range.start(), range.start() + half - 1),
            'B' => RangeInclusive::new(range.start() + half, *range.end()),
            _ => range
          }
        });

      let column = column_config
        .chars()
        .fold(RangeInclusive::new(0, 7), |range, letter| {
          let half = (range.end() + 1 - range.start()) / 2;

          match letter {
            'L' => RangeInclusive::new(*range.start(), range.start() + half - 1),
            'R' => RangeInclusive::new(range.start() + half, *range.end()),
            _ => range
          }
        });

      row.start() * 8 + column.start()
    })
}
